### ISSUE TITLE
[Blazor] Feature Request for .NET 7 - Enhanced Lazy Loading

### ISSUE THREAD

**Issue by daniel-p-tech** [dotnet/aspnetcore#39622]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

As it stands, lazy loading of modules has limitations that prevent it to be useful in real-world applications. We are designing a large enterprise app that is going to be comprised of 15-20 loosely coupled modules. Loading all modules during app initialization would cause delay during initial download of the application so it makes perfect sense to load modules dynamically.

I outlined problems with current implementation in a previous ticket.
https://github.com/dotnet/aspnetcore/issues/37829

While @javiercn did an excellent job (as always) suggesting a solution, I'm afraid this is more of a workaround that doesn't provide elegant developer experience.

### Describe the solution you'd like

1. I would like to keep all code required to run the module inside the module itself. There shouldn't be any shared libraries required to launch the module from the hosting app, this applies to DI initialization.

2. When module (assembly) is lazy loaded, allow execution of ``async`` initialization tasks inside some entry point method, maybe something similar to the proposal that was linked in the related issue:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/module-initializers
The caller should be notified when all initialization work is completed.

3. This entry point method should allow registration of additional DI services that are specific to the module (assembly).

4. Allow use of ``Inject`` attribute without any unnecessary ceremony.

5. Load dynamically any scoped CSS for the lazy loaded module. Currently there is only a single ``css`` file for all referenced assemblies residing in the hosted app, which of course is going to grow in size as the number of modules increases.

In short, I would like to see lazy loaded modules as completely isolated units of work that don't require any setup by the hosting app. The only necessary code inside the hosting app is to load the module and get a notification when the module is loaded (or failed to load).

Thank you for your consideration. I believe better mechanism for lazy loading of assemblies is critical to Blazor's adoption in large enterprise environments.

**Comment by javiercn** [comment]
@daniel-p-tech thanks for contacting us.

Unfortunately this is a problem that expands multiple areas and not something we plan to take on in the near future. It will require fundamental changes across different parts of the stack and we would need to see enough demand for it to justify such changes.

> I would like to keep all code required to run the module inside the module itself. There shouldn't be any shared libraries required to launch the module from the hosting app, this applies to DI initialization.

There are several challenges with this point. First of all, that it would require changes on the DI system to support instantiation of lazy dependencies. Currently the container is immutable once built, so its not possible to do this without a major (very possibly breaking) change. In addition to that, given that DI is a contract all different containers must implement, it would require all container implementations (not only the MS one) to follow suit and provide implementations for the functionality. While its not impossible, its very unlikely this happens. In any case I would suggest filing a separate ticket with your scenario and requirements in dotnet/runtime.

> When module (assembly) is lazy loaded, allow execution of async initialization tasks inside some entry point method, maybe something similar to the proposal that was linked in the related issue:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/module-initializers
The caller should be notified when all initialization work is completed.

This could be done in a few lines of code by your app when you load a dll, but since the point of this is to configure things like DI after the container is built, which is not possible, then it doesn't add value if you don't have point 1.

> 3\. This entry point method should allow registration of additional DI services that are specific to the module (assembly).

This is again, what I mentioned above.

> Allow use of Inject attribute without any unnecessary ceremony.

This would work naturally provided that the DI container supports the functionality/feature.

> Load dynamically any scoped CSS for the lazy loaded module. Currently there is only a single css file for all referenced assemblies residing in the hosted app, which of course is going to grow in size as the number of modules increases.

This is of limited value and something you can do today by referencing the `lib.scp.css` file for the module directly (if you look at the styles.css file you can see all the imports for other libraries). That said, this is not something we are likely going to do by default because the benefits are not worth the work and there is a way today to perform your own bundling.

I hope that you understand that while we appreciate your input we haven't seen a majority of our users require this level of sophistication and that as a result we are hesitant to add complexity to the framework without a clear signal that it will have a positive impact on the majority of our users.

That said, I would still suggest you file an issue in dotnet/runtime explaining the scenario for the DI requirements, as the core of your request relies on that. I would also recommend checking modular frameworks built for Blazor like [oqtane](https://www.oqtane.org/) as they might already provide what you are looking for.

Finally, it would be good for us if you can provide some data about how much bandwidth you think lazy loading is saving in your application. Most of the download cost comes from the runtime and the BCL which aren't lazy loaded, and app assemblies amount to a small portion of the download size, so it would be a good exercise to understand the concrete benefits on a particular scenario.

I understand that this might not be the answer you expect to receive, however I hope that at least this explanation helps clarify how we think about the topic.

**Comment by RyoukoKonpaku** [comment]
There are other DI containers that aren't immutable that allows late registrations. 
Did a quick experiment and the case is maybe possible without requesting changes from MEDI which as @javiercn stated would probably be hard due to reasons above.

For the quick experiment I used [DryIoc](https://github.com/dadhi/DryIoc) and it's adapter library `DryIoc.Microsoft.DependencyInjection` and replaced the MEDI as the container to use on Blazor WASM and placed the container onto a static class so it can be referenced on other places.

`Program.cs` looked like this.

```cs
var builder = WebAssemblyHostBuilder.CreateDefault(args);
var container = new Container(rules => rules.WithTrackingDisposableTransients()); // DryIoc container
builder.ConfigureContainer(new DryIocServiceProviderFactory(container), adaptedContainer =>
{
    StaticDependencyContainer.Container = adaptedContainer; // Assign it to a static class to use later
});

builder.RootComponents.Add<App>("#app");
builder.RootComponents.Add<HeadOutlet>("head::after");
builder.Services.AddScoped(sp => new HttpClient { BaseAddress = new Uri(builder.HostEnvironment.BaseAddress) });
await builder.Build().RunAsync();
```

I Confirmed it can dynamically register new services on the fly by having `Index.razor` have a button to register a service and `Counter.razor` load it via `@inject`. 

If I didn't click the button to register the service on `Index.razor`, navigating to Counter would throw an exception (stating the service doesn't exist) but if I did dynamically register the service on Index then Counter page loads fine and the service works fine. If you pair this with module initializers, I think it's possible to have isolated modules register their services on lazy load. I didn't test if module initializers work though, but if that works then it should be possible.

**Comment by javiercn** [comment]
@RyoukoKonpaku the issue is not that you can do it with one container framework, is that the platform needs to remain framework agnostic and hence the MEDI needs to support the features you are asking for.

If there are containers out there that allow you to register services after the fact, then the solution is not for us to build something, but for people who want this functionality to put the pieces together.

From our side, we need to keep compatible with the container abstractions in Microsoft.Extensions.DependencyInjection so any feature that we need, needs to be available there, and even if it were on the default implementation, we would be wary of taking a dependency on any feature that other containers don't have, as we don't want to limit the choice of DI container in the framework.

**Comment by RyoukoKonpaku** [comment]
@javiercn Completely understand the reasoning. The comment above was more on for @daniel-p-tech to see if there are solutions available with the current setup. Which seems there are, though it does mean using a different container other than MEDI for it.

**Comment by daniel-p-tech** [comment]
@RyoukoKonpaku thanks for the tip!

I'm not familiar with this library, are people successfully using it in real-world production applications? I published a bare-bones Blazor project plus DryIoc and the size of the two new Brotli compressed DLLs is about 240K which isn't exactly lightweight...

**Comment by RyoukoKonpaku** [comment]
> I'm not familiar with this library, are people successfully using it in real-world production applications?

We've been using it on a number of production apps right now. This library has always been my go to DI container if I needed more features that MEDI doesn't provide and the maintainer is pretty active. Though those apps are server apps, not Blazor WASM apps.

> I published a bare-bones Blazor project plus DryIoc and the size of the two new Brotli compressed DLLs is about 240K which isn't exactly lightweight...

Yeah this was one of my concerns when I thought up the quick experiment, the size increase by pulling another library might not be worth the trouble.
